🔀 Add `MergeModelCallBack`

August-murr opened the pull request:
# What does this PR do?
Fixes #2241

Since the focus was on replicating the checkpoint merging methods from [the paper](https://arxiv.org/abs/2410.10801), I have covered only Linear, TIES, SLERP, and DARE-TIES merging methods. These were the ones used in the paper and were primarily tested with the DPOTrainer.

Please provide feedback on any issues and suggest improvements regarding the structure of the files in this PR, as well as the documentation, to enhance clarity.

Here's an example of usage:

```bash
pip install trl[mergekit]
```
since it's an optional dependency

```python
from trl.mergekit_utils import MergeConfig
Trainer = DPOTrainer(...)
config = MergeConfig("ties")
config.target_model_path = "path_to_target_model" #if none is provided, the reference model will be used
config.policy_model_weight = 0.7 #optional
config.target_model_weight = 0.3 #optional 
merge_callback = MergeModelCallBack(config,push_to_hub=True,merge_at_every_checkpoint=True)
Trainer.add_callback(callback)
```



## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [x] Did you write any new necessary tests?


## Who can review?
@lewtun 
@qgallouedec

August-murr commented on setup.py:
```diff
@@ -103,6 +103,7 @@
     "deepspeed": ["deepspeed>=0.14.4"],
     "quantization": ["bitsandbytes<=0.41.1"],
     "llm_judge": ["openai>=1.23.2", "llm-blender>=0.0.2"],
+    "mergekit": ["mergekit @ git+https://github.com/arcee-ai/mergekit.git"],
```
This is not ideal and very messy, but it was the only way I could install mergekit from source since "pip" installed an older version

lewtun replied:
The `mergekit` maintainer has kindly made a new release. Would you mind setting that with a lower bound and seeing if the callback still works?

August-murr replied:
It works the same overall, but `mergekit` 0.0.5 uses an older version of `accelerate`, which breaks training unless you upgrade `accelerate` afterward. installing `mergekit` from source doesn't cause this problem

cg123 replied:
Sorry about that! 0.0.5.1 targets the current version of `accelerate`.

August-murr commented:
@lewtun 
@qgallouedec 
Feedback would be appreciated!

qgallouedec commented:
Thanks a lot @August-murr for the work. Can you add documentation, and test?

qgallouedec commented on trl/trainer/callbacks.py:
```diff
@@ -429,3 +430,59 @@ def on_step_end(self, args, state, control, **kwargs):
 
         # Save the last logged step, so we don't log the same completions multiple times
         self._last_logged_step = state.global_step
+class MergeModelCallBack(TrainerCallback):
```
```suggestion
class MergeModelCallback(TrainerCallback):
```

August-murr commented:
> Thanks a lot @August-murr for the work. Can you add documentation, and test?

I've already added most of the docs, as for the tests, unfortunately I won't be able to do it for a few days and if nobody else added them, I'll do it later.

August-murr commented:
The tests I added validate the success of the merge and I could expand it if necessary.
I also added docs to the callbacks file but was unable to produce the HTML file similar to the [callback docs](https://huggingface.co/docs/trl/main/en/callbacks) so I'd appreciate it if you could confirm whether the docs are properly generated or not.

lewtun commented on tests/test_callbacks.py:
```diff
@@ -221,3 +223,62 @@ def test_basic(self):
+    def setUp(self):
+        self.model = AutoModelForCausalLM.from_pretrained("trl-internal-testing/tiny-random-LlamaForCausalLM")
+        self.tokenizer = AutoTokenizer.from_pretrained("trl-internal-testing/tiny-random-LlamaForCausalLM")
+        self.dataset = load_dataset("trl-internal-testing/Anthropic-hh-rlhf-processed", split="train").select(range(8))
```
This dataset is a better choice to keep the tests running fast:

```suggestion
        self.dataset = load_dataset("trl-internal-testing/zen", "standard_preference", split="train")
```

lewtun commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,347 @@
+from huggingface_hub import HfApi
+
+
+def get_last_checkpoint_path(output_dir):
```
Can we reuse the `get_last_checkpoint()` [method](https://github.com/huggingface/transformers/blob/a3d69a8994d673899608a7c17fbf4f953f50474e/src/transformers/trainer_utils.py#L212) from `transformers` or is checking for the last modification time necessary here?

August-murr replied:
just replaced it

lewtun commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,347 @@
+    return last_checkpoint
+
+
+def upload_model_to_hf(folder_path, repo_id):
```
nit: let's add the types
```suggestion
def upload_model_to_hf(folder_path : str, repo_id: str):
```

lewtun commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,347 @@
+        repo_type=repo.repo_type,
+    )
+
+class MergeConfig:
```
Could you please add a docstring like we do for our model training configs: https://github.com/huggingface/trl/blob/623963126be5598bd1eea4ec82b43447fcc11535/trl/trainer/dpo_config.py#L34

This way it renders nicely in the docs :)

It would be good to provide some links to the various supported methods

qgallouedec replied:
Useful ressource: https://github.com/huggingface/trl/pull/1944#issue-2471984781

August-murr replied:
I got GPT to add the docstring. 
is it ok?

qgallouedec replied:
Sure!

lewtun commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,347 @@
+        else:
+            raise ValueError(f"Unknown merge method: {method}")
+    
+    def create_merge_config_linear(self):
```
```suggestion
    def create_merge_config_linear(self) -> MergeConfiguration:
```

lewtun commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,347 @@
+        """
+        Creates a merge configuration for a linear merge of two models with specified weights.
+        
+        Args:
```
Technically this method has no args, so I'd remove this part of the docstring

lewtun commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,347 @@
+            target_model_weight (float, optional): Weight for the target model. Defaults to 0.5.
+            dtype (str, optional): Data type for the merge. Defaults to 'float16'.
+        
+        Returns:
```
Can also be removed with the type hint above

lewtun commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,347 @@
+        
+        return merge_config
+
+    def create_merge_config_ties(self):
```
```suggestion
    def create_merge_config_ties(self) -> MergeConfiguration:
```

lewtun commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,347 @@
+        """
+        Creates a merge configuration for a TIES merge of two models, with specified weights and densities.
+        
+        Args:
```
Same as above: we can remove the args

lewtun commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,347 @@
+            normalize (float, optional): Normalization parameter. Defaults to 1.0.
+            dtype (str, optional): Data type for the merge. Defaults to 'float16'.
+        
+        Returns:
```
Same as above: we can remove this once we include the typehint

lewtun commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,347 @@
+        
+        return merge_config
+
+    def create_merge_config_dare_ties(self):
```
```suggestion
    def create_merge_config_dare_ties(self) -> MergeConfiguration:
```

lewtun commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,347 @@
+        """
+        Creates a merge configuration for a DARE TIES merge of two models, with specified weights and densities.
+        
+        Args:
```
Remove

lewtun commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,347 @@
+            normalize (float, optional): Normalization parameter. Defaults to 1.0.
+            dtype (str, optional): Data type for the merge. Defaults to 'float16'.
+        
+        Returns:
```
Remove

lewtun commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,347 @@
+        """
+        Creates a merge configuration for a SLERP merge of a model with a base model.
+        
+        Args:
```
Remove

lewtun commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,347 @@
+        
+        return merge_config
+
+    def create_merge_config_slerp(self):
```
```suggestion
    def create_merge_config_slerp(self) -> MergeConfiguration:
```

lewtun commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,347 @@
+            t_values (list, optional): List of ConditionalParameter values for SLERP. Defaults to None.
+            dtype (str, optional): Data type for the merge. Defaults to 'float16'.
+        
+        Returns:
```
Remove

lewtun commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,347 @@
+        return merge_config
+
+    
+    def create(self):
```
```suggestion
    def create(self) -> MergeConfiguration:
```

lewtun commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,347 @@
+            self.t_values = 0.5
+            self.dtype = 'float16'
+        else:
+            raise ValueError(f"Unknown merge method: {method}")
```
turbo nit:
```suggestion
            raise ValueError(f"Unsupported merge method: {method}")
```

lewtun commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,347 @@
+        elif self.method == 'slerp':
+            return self.create_merge_config_slerp()
+
+def Merge(config,out_path):
```
nit: we use snakecase for methods:

```suggestion
def merge_models(config : MergeConfig, out_path : str):
```

lewtun commented on trl/trainer/callbacks.py:
```diff
@@ -429,3 +430,59 @@ def on_step_end(self, args, state, control, **kwargs):
+            self.merge_config.policy_model_path = policy_model_path
+            if not self.merge_config.target_model_path:
+                self.merge_config.target_model_path = reference_model_path
+            output_path = policy_model_path+"/merged"
```
nit: we prefer f-strings in trl

```suggestion
            output_path = f"{policy_model_path}/merged"
```

lewtun commented on trl/trainer/callbacks.py:
```diff
@@ -429,3 +430,59 @@ def on_step_end(self, args, state, control, **kwargs):
+            if not self.merge_config.target_model_path:
+                self.merge_config.target_model_path = reference_model_path
+           
+            output_path = policy_model_path+"/merged"
```
```suggestion
            output_path = f"{policy_model_path}/merged"
```

lewtun reviewed (changes requested):
Thanks for adding this callback @August-murr - the implementation is very clean! I've left some minor comments on the PR and then I think it's good to merge (pun intended :))

Can you also fix the tests - it seems to be an issue with the path creation on Windows. Also can you add the callback to the docs here: https://github.com/huggingface/trl/blob/main/docs/source/callbacks.mdx

qgallouedec commented on trl/__init__.py:
```diff
@@ -88,10 +88,11 @@
         "SFTConfig",
         "SFTTrainer",
         "WinRateCallback",
+        "MergeModelCallback",
```
Please keep the alphabetical order

qgallouedec commented on trl/__init__.py:
```diff
@@ -88,10 +88,11 @@
         "XPOTrainer",
     ],
-    "trainer.callbacks": ["RichProgressCallback", "SyncRefModelCallback"],
+    "trainer.callbacks": ["RichProgressCallback", "SyncRefModelCallback","MergeModelCallback"],
```
```suggestion
    "trainer.callbacks": ["MergeModelCallback", "RichProgressCallback", "SyncRefModelCallback"],
```

qgallouedec commented on trl/__init__.py:
```diff
@@ -177,10 +178,11 @@
         XPOTrainer,
     )
-    from .trainer.callbacks import RichProgressCallback, SyncRefModelCallback
+    from .trainer.callbacks import RichProgressCallback, SyncRefModelCallback, MergeModelCallback
```
```suggestion
    from .trainer.callbacks import MergeModelCallback, RichProgressCallback, SyncRefModelCallback
```

qgallouedec commented on trl/__init__.py:
```diff
@@ -177,10 +178,11 @@
         SFTConfig,
         SFTTrainer,
         WinRateCallback,
+        MergeModelCallback,
```
Please keep the alphabetical order

qgallouedec commented on trl/mergekit_utils.py:
or just run `make precommit`, it will make it for you

qgallouedec commented on trl/trainer/__init__.py:
```diff
@@ -24,7 +24,7 @@
     "bco_config": ["BCOConfig"],
     "bco_trainer": ["BCOTrainer"],
-    "callbacks": ["LogCompletionsCallback", "RichProgressCallback", "SyncRefModelCallback", "WinRateCallback"],
+    "callbacks": ["LogCompletionsCallback", "RichProgressCallback", "SyncRefModelCallback", "WinRateCallback","MergeModelCallback"],
```
```suggestion
    "callbacks": ["LogCompletionsCallback", "MergeModelCallback", "RichProgressCallback", "SyncRefModelCallback", "WinRateCallback"],
```

qgallouedec commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,347 @@
+    )
+
+class MergeConfig:
+    def __init__(self, method: str):
```
Can we have a default value here?

qgallouedec commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,285 @@
+# Copyright 2022 The HuggingFace Team. All rights reserved.
```
```suggestion
# Copyright 2024 The HuggingFace Team. All rights reserved.
```

qgallouedec commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,285 @@
+    Args:
+        method (`str`, *optional*, defaults to `"linear"`):
+            The merge method to use. Supported methods include:
+            - `"linear"`: Linearly combines two models with specified weights.
```
```suggestion

            - `"linear"`: Linearly combines two models with specified weights.
```

otherwise it doesn't render properly

qgallouedec commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,285 @@
+            - `"ties"`: Combines two models using the TIES method with density parameters.
+            - `"dare_ties"`: A variant of TIES for domain adaptation.
+            - `"slerp"`: Combines models using spherical linear interpolation.
+    Note:
```
```suggestion

    Note:
```

qgallouedec commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,285 @@
+
+    Args:
+        method (`str`, *optional*, defaults to `"linear"`):
+            The merge method to use. Supported methods include:
```
```suggestion
            Merge method to use. Supported methods include:
```

qgallouedec commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,285 @@
+        t_values (`Optional[float]`): Interpolation factor for the SLERP method.
+        dtype (`str`): Data type to use for merging, e.g., `"float16"`.
+
+    Methods:
```
you can drop all of these

qgallouedec commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,285 @@
+        target_model_path (`Optional[str]`): Path to the target model.
+        policy_model_weight (`float`): Weight for the policy model (for `linear` and `ties` methods).
+        target_model_weight (`float`): Weight for the target model (for `linear` and `ties` methods).
+        policy_model_density (`Optional[List[float]]`): Density parameters for the policy model (for `ties` and `dare_ties`).
```
```suggestion
        policy_model_density (`List[float]`): Density parameters for the policy model (for `ties` and `dare_ties`).
```

They are not optional per se. See typing doc (same for the other)

qgallouedec commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,285 @@
+        if method == "linear":
+            self.policy_model_weight = 0.5
+            self.target_model_weight = 0.5
+            self.dtype = "float16"
```
Why do you need this?

August-murr replied:
default values taken from examples in the Mergekits repo.
can be set or changed by the user

qgallouedec replied:
the default from the class definition is `None` (see https://github.com/arcee-ai/mergekit/blob/57e7d14e2a732f532970e2c9dada00e2d8f15a7a/mergekit/config.py#L91) so I would keep it like this

qgallouedec commented on trl/mergekit_utils.py:
```diff
@@ -0,0 +1,285 @@
+                {
+                    "model": {
+                        "model": {"path": self.policy_model_path, "revision": None},
+                        "lora": None,
```
Can we omit the values that are none?

August-murr replied:
I left the values as `None` so we can add support for LoRA or other MergeKit features in the future if needed.

qgallouedec commented on trl/trainer/callbacks.py:
```diff
@@ -429,3 +430,64 @@ def on_step_end(self, args, state, control, **kwargs):
+    """
+
+    def __init__(self, merge_config, push_to_hub=False, merge_at_every_checkpoint=False):
+        self.push_to_hub = push_to_hub
```
push_to_hub should be gotten from the training args  I think

August-murr replied:
push_to_hub from training args pushes the trained model 
this one pushes the merged model

lewtun reviewed (approved):
Thanks for iterating @August-murr ! The PR LGTM now and once the CI is green & @qgallouedec approves, I think we can merge it

August-murr commented:
> Thanks for iterating @August-murr ! The PR LGTM now and once the CI is green & @qgallouedec approves, I think we can merge it

The tests without optional dependency failed since Mergekit is an optional dependency

kashif commented:
@August-murr in the `import_utils` you can define a new `is_mergekit_available` helper and then in the tests you can skip the tests if its not available

qgallouedec commented:
Like here:

https://github.com/huggingface/trl/blob/6f8fe59aebc1153c6000c922b8edc4bb11efd506/trl/import_utils.py#L39-L40

https://github.com/huggingface/trl/blob/6f8fe59aebc1153c6000c922b8edc4bb11efd506/tests/testing_utils.py#L42-L46

https://github.com/huggingface/trl/blob/6f8fe59aebc1153c6000c922b8edc4bb11efd506/tests/test_judges.py#L62-L63

Don't hesitate to ask for help if you want the maintainers to do it for you.

August-murr commented:
> Like here:
> 
> https://github.com/huggingface/trl/blob/6f8fe59aebc1153c6000c922b8edc4bb11efd506/trl/import_utils.py#L39-L40
> 
> https://github.com/huggingface/trl/blob/6f8fe59aebc1153c6000c922b8edc4bb11efd506/tests/testing_utils.py#L42-L46
> 
> https://github.com/huggingface/trl/blob/6f8fe59aebc1153c6000c922b8edc4bb11efd506/tests/test_judges.py#L62-L63
> 
> Don't hesitate to ask for help if you want the maintainers to do it for you.

Done!

qgallouedec commented on tests/testing_utils.py:
```diff
@@ -44,3 +45,10 @@ def require_llm_blender(test_case):
+
+def require_mergekit(test_case):
+    """
+    Decorator marking a test that requires Mergekit. Skips the test if Mergekit is not available.
+    """
+    return unittest.skipUnless(is_mergekit_available(), "test requires Mergekit")(test_case)
```
```suggestion
    Decorator marking a test that requires Mergekit. Skips the test if mergekit is not available.
    """
    return unittest.skipUnless(is_mergekit_available(), "test requires mergekit")(test_case)
```

qgallouedec commented on trl/__init__.py:
```diff
@@ -88,10 +88,11 @@
         "XPOTrainer",
     ],
-    "trainer.callbacks": ["RichProgressCallback", "SyncRefModelCallback"],
+    "trainer.callbacks": ["RichProgressCallback", "SyncRefModelCallback", "MergeModelCallback"],
```
```suggestion
    "trainer.callbacks": ["MergeModelCallback", "RichProgressCallback", "SyncRefModelCallback"],
```

qgallouedec commented:
Nice, thanks! Just running some tests, waiting for the CI to be green, and we're good to merge (expect some commits from me on this branch)

qgallouedec commented:
Another question that raised while reviewing: while having a new configuration class while we can directly use the one from `mergekit`. I'm afraid that I could confuse the user, tempted to use:

```python
from mergekit import MergeConfiguration
from trl import MergeModelCallback
    
merge_callback = MergeModelCallback(MergeConfiguration())
```

August-murr commented:
> Another question that came up during the review: why have a new configuration class when we can use the mergekit one directly? I'm afraid of confusing the user, tempted to use :
> 
> ```python
> from mergekit import MergeConfiguration
> from trl import MergeModelCallback
>     
> merge_callback = MergeModelCallback(MergeConfiguration())
> ```

Actually, ease of use for the user was the reason why I had to write the class in mergekit_utils since mergekit uses a yaml file to get it's Merge config, which is easier to implement but more complicated for the user.

and if you wanted to use `MergeConfiguration` directly from mergekit:
```python
from mergekit.config import MergeConfiguration

merge_config_dict = {
    "dtype": "float16",
    "merge_method": "linear",
    "models": [
        {"model": "path_to_model_1", "parameters": {"weight": 0.4}},
        {"model": "path_to_model_2", "parameters": {"weight": 0.6}},
    ],
}

config = MergeConfiguration.model_validate(merge_config_dict)
```
As you add more parameters to the configuration, the dictionary becomes increasingly nested.

The current implementation, although harder to maintain, simplifies everything for the user:
```python
from trl.mergekit_utils import MergeConfig
config = MergeConfig("linear")
config.policy_model_weight = 0.4
config.target_model_weight = 0.6
```

qgallouedec commented:
That makes sense.
Do you think we can get the best of both worlds by making `trl.MergeConfig` inherits from `mergekit.config.MergeConfigurationMergeConfig`?

August-murr commented:
> That makes sense.
> Do you think we can get the best of both worlds by making `trl.MergeConfig` inherits from `mergekit.config.MergeConfigurationMergeConfig`?

I'll figure it out.

qgallouedec commented:
LGTM thanks!
I've just applied some minor refinements:

- compat with windows file path
- sort imports and function
- common method for saving and pushing in the callback
- add "trl" to model tags

August-murr commented:
@qgallouedec 
About the failed tests:
The tests do not fail on Ubuntu; they only fail on Windows. I realized that the issue arose from a permission error from the temporary directory when trying to delete the merged files, specifically the `model.safetensors.`

qgallouedec commented:
> @qgallouedec About the failed tests: The tests do not fail on Ubuntu; they only fail on Windows. I realized that the issue arose from a permission error from the temporary directory when trying to delete the merged files, specifically the `model.safetensors.`

Ah thanks, I was debugging, but I don't have access to windows vm right now (explains https://github.com/huggingface/trl/pull/2282/commits/fa5bafe617793ed340303cf0ebded6ac03cab39f). Any idea how to solve it?

qgallouedec commented:
Found a solution with a57d88a1b317785fa85e3b09bd463ecb0b9eef06

August-murr commented:
@qgallouedec 
Sorry I wasn't able to sort it out myself.
